uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h

```diff
@@ -0,0 +1,794 @@
+  /// how many bytes it's actually written, we can write the length prefix, and
+  /// advance the current offset.
+  ///
+  /// Safety: The closure must copy the string's contents into the span, and
+  /// return the exact number of bytes it copied. Returning the wrong count can
+  /// either truncate the string, or leave uninitialized memory in the buffer.
+  /// The closure must not call any other methods on the writer.
+  void WriteRawString(size_t aSizeHint,
+                      const std::function<size_t(Span<char>)>& aClosure) {
+    // First, make sure the buffer is big enough to hold the length prefix.
+    // We'll start writing our string directly after the prefix.
+    CheckedInt<size_t> newOffset = mOffset;
+    newOffset += sizeof(uint32_t);
+    AssertInBounds(newOffset);
+    char* begin =
+        reinterpret_cast<char*>(&mBuffer.Elements()[newOffset.value()]);
+
+    // Next, ensure the buffer has space for enough bytes up to the size hint.
+    // We may write fewer bytes than hinted, but we need to handle the worst
+    // case if needed.
+    newOffset += aSizeHint;
+    AssertInBounds(newOffset);
+
+    // Call the closure to write the bytes directly into the buffer.
+    size_t bytesWritten = aClosure(Span(begin, aSizeHint));
+
+    // Great, now we know the real length! Write it at the beginning.
+    uint32_t lengthPrefix = PR_htonl(bytesWritten);
+    memcpy(&mBuffer.Elements()[mOffset.value()], &lengthPrefix,
+           sizeof(uint32_t));
+
+    // And figure out our actual offset.
+    newOffset -= aSizeHint;
+    newOffset += bytesWritten;
+    AssertInBounds(newOffset);
+    mOffset = newOffset;
+  }
+
+  RustBuffer ToRustBuffer() {
+    auto size = static_cast<uint32_t>(mOffset.value());
+    auto buffer = {{ ci.ffi_bytebuffer_alloc().name() }}(size);
+    memcpy(buffer.mData, mBuffer.Elements(), size);
+    return buffer;
+  }
+
+ private:
+  void AssertInBounds(const CheckedInt<size_t>& aNewOffset) const {
+    MOZ_RELEASE_ASSERT(aNewOffset.isValid() &&
+                       aNewOffset.value() <= mBuffer.Capacity());
+  }
+
+  template <typename T>
+  void WriteAt(const T& aValue,
+               const std::function<void(size_t, const T&)>& aClosure) {
+    CheckedInt<size_t> newOffset = mOffset;
+    newOffset += sizeof(T);
+    AssertInBounds(newOffset);
+    mBuffer.SetLength(newOffset.value());
+    aClosure(mOffset.value(), aValue);
+    mOffset = newOffset;
+  }
+
+  nsTArray<uint8_t> mBuffer;
+  CheckedInt<size_t> mOffset;
+};
+
+/// A "trait" struct with specializations for types that can be read and
+/// written into a byte buffer. This struct is specialized for all serializable
+/// types.
+template <typename T>
+struct Serializable {
+  /// Returns the size of the serialized value, in bytes. This is used to
+  /// calculate the allocation size for the Rust byte buffer.
+  static CheckedInt<size_t> Size(const T& aValue) = delete;
```

Comment:
Why is the return type `CheckedInt`?

Reply:
I was using `CheckedInt` throughout to avoid accidental overflows when serializing, but with the `RustBuffer` changes on `main` now, `Size` isn't necessary anymore. I think we can remove `CheckedInt` from the other uses, too, since `RustBuffer`s can only hold an `int32_t` worth of bytes.